File: Supplementary_webpages/static/gallery_res/Convex Optimization page 154/convex_optimization_154.py
"""
∑_i f_i²p_i - (∑_i f_i p_i)²

where

f_i ∈ ℝ  
p_i ∈ ℝ  
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


class convex_optimization_154ResultType:
    def __init__( self, ret):
        self.ret = ret


def convex_optimization_154(f, p):
    f = np.asarray(f, dtype=np.float64)
    p = np.asarray(p, dtype=np.float64)

    dim_0 = f.shape[0]
    assert f.shape == (dim_0,)
    assert p.shape == (dim_0,)

    sum_0 = 0
    for i in range(1, len(p)+1):
        sum_0 += np.power(f[i-1], 2) * p[i-1]
    sum_1 = 0
    for i in range(1, len(p)+1):
        sum_1 += f[i-1] * p[i-1]
    ret = sum_0 - np.power((sum_1), 2)
    return convex_optimization_154ResultType(ret)


def generateRandomData():
    dim_0 = np.random.randint(10)
    f = np.random.randn(dim_0)
    p = np.random.randn(dim_0)
    return f, p


if __name__ == '__main__':
    f, p = generateRandomData()
    print("f:", f)
    print("p:", p)
    func_value = convex_optimization_154(f, p)
    print("return value: ", func_value.ret)

File: Supplementary_webpages/static/gallery_res/Convex Optimization page 154/convex_optimization_154.cpp
/*
∑_i f_i²p_i - (∑_i f_i p_i)²

where

f_i ∈ ℝ  
p_i ∈ ℝ  
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

struct convex_optimization_154ResultType {
    double ret;
    convex_optimization_154ResultType(const double & ret)
    : ret(ret)
    {}
};

convex_optimization_154ResultType convex_optimization_154(
    const std::vector<double> & f,
    const std::vector<double> & p)
{
    const long dim_0 = f.size();
    assert( p.size() == dim_0 );

    double sum_0 = 0;
    for(int i=1; i<=p.size(); i++){
        sum_0 += pow(f.at(i-1), 2) * p.at(i-1);
    }
    double sum_1 = 0;
    for(int i=1; i<=p.size(); i++){
        sum_1 += f.at(i-1) * p.at(i-1);
    }
    double ret = sum_0 - pow((sum_1), 2);
    return convex_optimization_154ResultType(ret);
}


void generateRandomData(std::vector<double> & f,
    std::vector<double> & p)
{
    const int dim_0 = rand()%10;
    f.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        f[i] = rand() % 10;
    }
    p.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        p[i] = rand() % 10;
    }
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    std::vector<double> f;
    std::vector<double> p;
    generateRandomData(f, p);
    convex_optimization_154ResultType func_value = convex_optimization_154(f, p);
    std::cout<<"return value:\n"<<func_value.ret<<std::endl;
    return 0;
}